ttnn/cpp/ttnn/operations/moreh/moreh_linear_backward/moreh_linear_backward.cpp

```diff
@@ -0,0 +1,186 @@
+    const std::optional<ttnn::MemoryConfig>& weight_grad_mem_config,
+    const std::optional<ttnn::MemoryConfig>& bias_grad_mem_config,
+    const DeviceComputeKernelConfig compute_kernel_config) {
+    std::vector<std::optional<Tensor>> result(3);
+    const auto [input_required_grad, weight_required_grad, bias_required_grad] =
+        get_required_outputs(are_required_outputs);
+
+    TT_ASSERT(
+        output_grad.storage_type() == StorageType::DEVICE && input.storage_type() == StorageType::DEVICE &&
+            weight.storage_type() == StorageType::DEVICE,
+        "input and weight tensors need to be on device");
+
+    TT_FATAL(output_grad.storage_type() == StorageType::DEVICE, "Error");
+    auto kernel_config_val =
+        init_device_compute_kernel_config(output_grad.device()->arch(), compute_kernel_config, MathFidelity::HiFi4);
+    moreh_linear_backward_validate(output_grad, input, weight, input_grad, weight_grad, bias_grad);
+
+    if (input_required_grad) {
+        TT_ASSERT(input_grad.has_value(), "input_grad tensor should not be std::nullopt");
+        result[0] = ttnn::moreh_matmul(
+            output_grad, weight, false, false, input_grad, std::nullopt, input_grad_mem_config, kernel_config_val);
+    }
+
+    if (weight_required_grad) {
+        TT_ASSERT(weight_grad.has_value(), "weight_grad tensor should not be std::nullopt");
+        const auto& weight_grad_tensor = weight_grad.value();
+
+        if (is_same_batch_dim(output_grad, weight_grad_tensor)) {
+            ttnn::moreh_matmul(
+                output_grad,
+                input,
+                true,
+                false,
+                weight_grad_tensor,
+                std::nullopt,
+                weight_grad_mem_config,
+                kernel_config_val);
+        } else {
+            const auto& temp_weight_grad = ttnn::moreh_matmul(
+                output_grad, input, true, false, std::nullopt, std::nullopt, weight_grad_mem_config, kernel_config_val);
+            TT_ASSERT(weight_grad.has_value(), "weight_grad tensor should not be std::nullopt");
+            std::vector<int64_t> dims =
+                find_reduce_dim(temp_weight_grad.get_legacy_shape(), weight_grad.value().get_legacy_shape());
+            ttnn::moreh_sum(
+                temp_weight_grad, dims, true, weight_grad.value(), weight_grad_mem_config, kernel_config_val);
+        }
+        result[1] = weight_grad_tensor;
+    }
+
+    if (bias_required_grad) {
+        std::vector<std::optional<Tensor>> output_tensors = ttnn::prim::moreh_linear_backward(
+            output_grad,
+            input,
+            weight,
+            are_required_outputs,
+            bias,
+            input_grad,
+            weight_grad,
+            bias_grad,
+            input_grad_mem_config,
+            weight_grad_mem_config,
+            bias_grad_mem_config,
+            kernel_config_val);
+        result[2] = std::make_optional(output_tensors.at(0).value());
```

Comment:
This part seems like it should only calculate bias_grad, but it looks strange to be calling `moreh_linear_backward`.

Reply:
resolved in https://github.com/tenstorrent/tt-metal/pull/12812/commits/8dec1ff696bd9948f9366cbee8ec5733a58f2081